[API Compatibility] Add pir support for SDPA and set enable_gqa default to True for better compatibility

LittleHeroZZZX opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->
User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->
Bug fixes

### Description
<!-- Describe what you’ve done -->
本 PR 修复了 SDPA 在静态图下报错问题和 https://github.com/PaddlePaddle/Paddle/pull/76446 引入的不兼容升级，具体包括：
- 静态图下使用 `paddle.framework._current_expected_place_()` 作为 qkv 的place 选择 SDPA 后端
- 根据 `paddle.device.is_bf16_supported(including_emulation=False)` 判断当前 cuda device 是否支持 BF16
- `paddle.nn.F.SDPA` 参数 `enable_gqa` 默认为 True，以对齐升级前 SDPA 的行为
- 当 `enable_gqa` 为 True 时，后置 repeat_head 的行为，对于支持 GQA 的后端直接透传 QKV
- 支持 0-size Tensor 输入

card-94386

LittleHeroZZZX committed:
Set enable_GQA default to True for F.SDPA

LittleHeroZZZX committed:
feat: support static graph for SDPA  and improve GQA handling

- Add paddle.pir.Value to SUPPORTED_ATTN_BIAS_TYPES and _get_tensor_bias check
- Extend softmax_mask_fuse_upper_triangle condition to exclude CPU and non-square matrices
- Refactor bfloat16 support detection in SDPA config initialization
- Simplify GQA expansion logic by removing manual tensor expansion for repeated heads

LittleHeroZZZX committed:
feat(nn/functional): correct GQA acronym expansion in docstring

Changed the expansion of the GQA acronym from "Generic Query Attention" to "Group Query Attention" in the scaled_dot_product_attention function docstring. This ensures the documentation accurately reflects the terminology used for the attention mechanism.

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -555,6 +558,11 @@ def scaled_dot_product_attention(
 
+    qkv_place = (query.place, key.place, value.place)
+    if not paddle.base.in_dygraph_mode():
+        with paddle.base.dygraph.guard():
```
静态图下还能用这个动态图guard吗

LittleHeroZZZX replied:
这 guard 就是为了在静态图下创建出一份动态图的上下文，实测没问题

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -508,32 +520,23 @@ def scaled_dot_product_attention(
-            )
-
+    else:
+        assert q_heads == k_heads == v_heads, (
```
确认一下，torch的enable_gqa为False时，针对q/k/v的shape不同时是否也是抛出异常处理

LittleHeroZZZX replied:
是的
<img width="1944" height="376" alt="image" src="https://github.com/user-attachments/assets/0e55c44c-be3e-4347-a215-dd642f124dc9" />

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -41,22 +41,34 @@
-            else (paddle.float16, paddle.float),
-        },
-    }
+    with paddle.base.dygraph.guard():
```
这些guard的功能是啥

LittleHeroZZZX replied:
静态图 下 `paddle.framework._current_expected_place()` 无法获取当前的默认 device

zhwesky2010 replied:
> 静态图 下 `paddle.framework._current_expected_place()` 无法获取当前的默认 device

直接使用`Tensor.is_cuda` 这个接口吧，动静态图下都封装好了

<img width="581" height="252" alt="infoflow 2025-12-02 13-50-08" src="https://github.com/user-attachments/assets/11f80774-206b-4a70-9a7e-da1eefda5c04" />

LittleHeroZZZX replied:
这里是在初始化一些配置，不应该依赖传入的 Tensor。另外还要考虑 CustomDevice 是否支持 bf16，不仅仅是 cuda

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -41,22 +41,34 @@
+        is_bf16_natively_supported = False
+        if paddle.base.is_compiled_with_cuda():
+            is_bf16_natively_supported = (
+                paddle.cuda.get_device_capability()[0] >= 8
```
这个是我们之前的is_bf16_supported计算得不准确吗，这些逻辑是否加到`is_bf16_supported` 里去

LittleHeroZZZX replied:
不是，`is_bf16_supported` 表示框架在当前硬件上是否支持 bf16，而不是当前硬件是否原生支持 bf16（这点与torch 对齐）。尽管在 V100 上框架支持 bf16，但是由于非硬件支持， SDPA 相关的kernel 基本都没注册

zhwesky2010 replied:
> 不是，`is_bf16_supported` 表示框架在当前硬件上是否支持 bf16，而不是当前硬件是否原生支持 bf16（这点与torch 对齐）。尽管在 V100 上框架支持 bf16，但是由于非硬件支持， SDPA 相关的kernel 基本都没注册

是说下面的check_all_tensors_on_device，直接用`Tensor.is_cuda`吧，不用搞得太复杂了。

这里是另一个问题，is_bf16_natively_supported是为了解决对不齐torch的问题？那`paddle.device.is_bf16_supported()`能与torch对齐吗

LittleHeroZZZX replied:
> 是说下面的check_all_tensors_on_device，直接用`Tensor.is_cuda`吧，不用搞得太复杂了。

我刚发现也不用这些动态图的 guard ，静态图下 `paddle.framework._current_expected_place_()` 能拿到正确的device，之前调的没有最后下划线需要在动态图的上下文里才能拿到 device

>这里是另一个问题，is_bf16_natively_supported是为了解决对不齐torch的问题？

既是跟torch 对齐，也是适配我们 paddle 自己。 MEA 在V100 (SM<80) 的机器上是不支持 BF16的，不能根据 is_bf16_supported 来判断是否支持 BF16，而是要根据 SM

>那paddle.device.is_bf16_supported()能与torch对齐吗
与torch 一致，在 SM <80 的机器上返回 True

LittleHeroZZZX replied:
> 是说下面的check_all_tensors_on_device，直接用Tensor.is_cuda吧

这个反而没必要吧，现在整个用来选择后端的 Param 里没有把Tensor 传进来，而是直接根据元数据来选择的

zhwesky2010 replied:
> > 是说下面的check_all_tensors_on_device，直接用Tensor.is_cuda吧
> 
> 这个反而没必要吧，现在整个用来选择后端的 Param 里没有把Tensor 传进来，而是直接根据元数据来选择的

OK，直接判断place吧。

zhwesky2010 replied:
> > 是说下面的check_all_tensors_on_device，直接用`Tensor.is_cuda`吧，不用搞得太复杂了。
> 
> 我刚发现也不用这些动态图的 guard ，静态图下 `paddle.framework._current_expected_place_()` 能拿到正确的device，之前调的没有最后下划线需要在动态图的上下文里才能拿到 device
> 
> > 这里是另一个问题，is_bf16_natively_supported是为了解决对不齐torch的问题？
> 
> 既是跟torch 对齐，也是适配我们 paddle 自己。 MEA 在V100 (SM<80) 的机器上是不支持 BF16的，不能根据 is_bf16_supported 来判断是否支持 BF16，而是要根据 SM
> 
> > 那paddle.device.is_bf16_supported()能与torch对齐吗
> > 与torch 一致，在 SM <80 的机器上返回 True

这块逻辑如果适合插入到device.is_bf16_supported就插入进去吧。

不适合的话，就针对is_bf16_supported做一个定制的函数，这里搞一些动态图guard挺奇怪的，静态图下能用吗（动态图guard一般就是专门针对动态图的，虽然这里是为了一些特殊的配置）

zhwesky2010 commented on python/paddle/nn/functional/flash_attention.py:
```diff
@@ -148,6 +148,8 @@ def _math_attention(
             "xpu" in place
             or product.shape[-1] < 32
             or product.shape[-1] > 16384
+            or place == paddle.CPUPlace()
```
这里修改的原因是？

LittleHeroZZZX replied:
这个分支内的 softmax_mask_fuse_upper_triangle
不支持 CPU、不支持 src_seq_len != target_seq_len的情况

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -625,6 +646,18 @@ def scaled_dot_product_attention(
 
     elif sdp_func_name == "math":
+        # repeat_interleave does not support float16 on GPU, so we manually expand the tensor
+        if k_heads != q_heads:
```
加一个_repeat_kv的方法，复用下这两个重复代码吧

LittleHeroZZZX replied:
Done

LittleHeroZZZX committed:
Add 0size input support for sdpa

LittleHeroZZZX committed:
feat(sdpa): extract key-value repetition logic into helper function

Refactor the scaled_dot_product_attention function by extracting the duplicate key-value tensor repetition logic into a new helper function _repeat_kv. This improves code maintainability by eliminating redundancy and centralizing the workaround for float16 GPU limitations with repeat_interleave. The helper function handles the expansion of key and value tensors along the num_heads dimension when the number of query heads exceeds key/value heads.

LittleHeroZZZX committed:
feat(test): skip zero-size attention test on CPU due to fp16 matmul unsupported

Add skip decorator to TestZeroSizeBase class to prevent execution on CPU environments where fp16 matrix multiplication is not supported. This ensures tests only run on compatible hardware, avoiding failures and improving test reliability.

LittleHeroZZZX committed:
feat(sdpa): remove unnecessary guard context for static graph mode

Remove the `with paddle.base.dygraph.guard():` context when not in dygraph mode, as it is unnecessary for static graph execution. Directly assign `qkv_place` using `_current_expected_place_()` to simplify the code and avoid potential overhead from the guard context.

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -581,9 +615,14 @@ def scaled_dot_product_attention(
-        mask_shape = (bs, num_heads_q, seq_len_q, seq_len_k)
-        attn_mask = attn_mask.expand(mask_shape)
-    sdp_func_name = select_sdp_for_sdpa(param)
+        if attn_mask.shape.numel() != 0:
```
这个cornel case导致插入了很多额外判断进来，有这个用法吗。是不是至多只有q/k/v可能是0-size

LittleHeroZZZX replied:
这个是 QA 监控到这样的 case 挂了后补的， attnmask 含有 0-Size 语义上确实没问题

LittleHeroZZZX committed:
feat(device): update bf16 support check and SDPA configuration

- Fix function call from `_current_expected_place` to `_current_expected_place_` in `is_bf16_supported`
- Simplify `init_config` in SDPA module by directly using `paddle.device.is_bf16_supported` with `including_emulation=False`
- Remove redundant CUDA capability check and streamline configuration logic for flash and memory-efficient attention

zhwesky2010 reviewed (approved):
LGTM

LittleHeroZZZX commented:
/re-run approval
